File: LogisimPlus/LogisimPlus/draw/actions/ModelChangeAttributeAction.cs
// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// To produce customized conversions, purchase a Premium Edition license:
// https://www.tangiblesoftwaresolutions.com/product-details/java-to-csharp-converter.html
// ====================================================================================================

using System.Collections.Generic;

/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

namespace draw.actions
{

	using AttributeMapKey = draw.model.AttributeMapKey;
	using CanvasModel = draw.model.CanvasModel;
	using CanvasObject = draw.model.CanvasObject;
	using logisim.data;

	public class ModelChangeAttributeAction : ModelAction
	{
		private Dictionary<AttributeMapKey, object> oldValues;
		private Dictionary<AttributeMapKey, object> newValues;
// JAVA TO C# CONVERTER WARNING: Java wildcard generics have no direct equivalent in C#:
// ORIGINAL LINE: private logisim.data.Attribute<?> attr;
		private Attribute attr;

		public ModelChangeAttributeAction(CanvasModel model, Dictionary<AttributeMapKey, object> oldValues, Dictionary<AttributeMapKey, object> newValues) : base(model)
		{
			this.oldValues = oldValues;
			this.newValues = newValues;
		}

		public override ICollection<CanvasObject> Objects
		{
			get
			{
				HashSet<CanvasObject> ret = new HashSet<CanvasObject>();
				foreach (AttributeMapKey key in newValues.Keys)
				{
					ret.Add(key.Object);
				}
				return ret;
			}
		}

		public override string Name
		{
			get
			{
	// JAVA TO C# CONVERTER WARNING: Java wildcard generics have no direct equivalent in C#:
	// ORIGINAL LINE: logisim.data.Attribute<?> a = attr;
				Attribute a = attr;
				if (a == null)
				{
					bool found = false;
					foreach (AttributeMapKey key in newValues.Keys)
					{
	// JAVA TO C# CONVERTER WARNING: Java wildcard generics have no direct equivalent in C#:
	// ORIGINAL LINE: logisim.data.Attribute<?> at = key.getAttribute();
						Attribute at = key.Attribute;
						if (found)
						{
							if (a == null ? at != null :!a.Equals(at))
							{
								a = null;
								break;
							}
						}
						else
						{
							found = true;
							a = at;
						}
					}
					attr = a;
				}
				if (a == null)
				{
					return Strings.get("actionChangeAttributes");
				}
				else
				{
					return Strings.get("actionChangeAttribute", a.DisplayName);
				}
			}
		}

		internal override void doSub(CanvasModel model)
		{
			model.AttributeValues = newValues;
		}

		internal override void undoSub(CanvasModel model)
		{
			model.AttributeValues = oldValues;
		}
	}

}

File: logisim/app/src/main/java/draw/actions/ModelChangeAttributeAction.java
/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

package draw.actions;

import java.util.Collection;
import java.util.HashSet;
import java.util.Map;
import java.util.Objects;

import draw.model.AttributeMapKey;
import draw.model.CanvasModel;
import draw.model.CanvasObject;
import logisim.data.Attribute;

public class ModelChangeAttributeAction extends ModelAction {
	private Map<AttributeMapKey, Object> oldValues;
	private Map<AttributeMapKey, Object> newValues;
	private Attribute<?> attr;

	public ModelChangeAttributeAction(CanvasModel model, Map<AttributeMapKey, Object> oldValues,
			Map<AttributeMapKey, Object> newValues) {
		super(model);
		this.oldValues = oldValues;
		this.newValues = newValues;
	}

	@Override
	public Collection<CanvasObject> getObjects() {
		HashSet<CanvasObject> ret = new HashSet<>();
		for (AttributeMapKey key : newValues.keySet()) ret.add(key.getObject());
		return ret;
	}

	@Override
	public String getName() {
		Attribute<?> a = attr;
		if (a == null) {
			boolean found = false;
			for (AttributeMapKey key : newValues.keySet()) {
				Attribute<?> at = key.getAttribute();
				if (found) {
					if (!Objects.equals(a, at)) {
						a = null;
						break;
					}
				} else {
					found = true;
					a = at;
				}
			}
			attr = a;
		}
		return a == null ? Strings.get("actionChangeAttributes") : Strings.get("actionChangeAttribute", a.getDisplayName());
	}

	@Override
	void doSub(CanvasModel model) {
		model.setAttributeValues(newValues);
	}

	@Override
	void undoSub(CanvasModel model) {
		model.setAttributeValues(oldValues);
	}
}
